File: java_maven/src/main/java/edu/usc/ict/superglu/core/blackwhitelist/BlackWhiteListEntry.java
package edu.usc.ict.superglu.core.blackwhitelist;

import edu.usc.ict.superglu.core.BaseMessage;
import edu.usc.ict.superglu.core.GIFTMessage;
import edu.usc.ict.superglu.core.Message;
import edu.usc.ict.superglu.core.VHMessage;
import edu.usc.ict.superglu.util.SerializationConvenience;
import edu.usc.ict.superglu.util.StorageToken;
import edu.usc.ict.superglu.util.SuperGlu_Serializable;

/**
 * This class represents a single entry in the black or white lists.
 * 
 * The each of these fields can be filled with a value or a wildcard (*)
 * 
 * @author auerbach
 *
 */

public class BlackWhiteListEntry extends SuperGlu_Serializable {

	private static final String MESSAGE_TYPE_KEY = "messageType";
	private static final String VERSION_KEY = "version";
	private static final String MESSAGE_NAME_KEY = "messageName";
	
	private static final String WILDCARD = "*";
	
	public static final String GIFT_MESSAGE_TYPE = "GIFT";
	public static final String SUPERGLU_MESSAGE_TYPE = "SuperGLU";
	public static final String VHUMAN_MESSAGE_TYPE = "VHuman";
	
	

	private String messageType;
	private String version;
	private String messageName;

	public BlackWhiteListEntry() {
		this.messageType = null;
		this.version = null;
		this.messageName = null;
	}

	public BlackWhiteListEntry(String messageType, String version, String messageName) {
		this.messageType = messageType;
		this.messageName = messageName;
		this.version = version;
	}

	public BlackWhiteListEntry(String entryAsString) {
		String[] tokenizedString = entryAsString.split("\\.");

		if (tokenizedString.length < 3) {
			logger.error("Failed to parse black/white list entry: " + entryAsString);
			throw new RuntimeException("failed to parse black/white list entry :" + entryAsString);
		}

		this.messageType = tokenizedString[0];
		this.version = tokenizedString[1];
		this.messageName = tokenizedString[2];
	}

	@Override
	public boolean equals(Object otherObject) {
		if (!super.equals(otherObject))
			return false;

		BlackWhiteListEntry other = (BlackWhiteListEntry) otherObject;

		if (!fieldIsEqual(this.messageName, other.messageName))
			return false;

		if (!fieldIsEqual(this.messageType, other.messageType))
			return false;

		if (!fieldIsEqual(this.version, other.version))
			return false;

		return true;
	}

	@Override
	public int hashCode() {
		int result = super.hashCode();
		int arbitraryPrimeNumber = 23;

		if (this.messageName != null)
			result = result * arbitraryPrimeNumber + this.messageName.hashCode();

		if (this.messageType != null)
			result = result * arbitraryPrimeNumber + this.messageType.hashCode();

		if (this.version != null)
			result = result * arbitraryPrimeNumber + this.version.hashCode();

		return result;
	}
	
	

	@Override
	public String toString() {
		return this.messageType + "." + this.version + "." + this.messageName;
	}

	@Override
	public void initializeFromToken(StorageToken token) {
		super.initializeFromToken(token);

		this.messageName = (String) SerializationConvenience.untokenizeObject(token.getItem(MESSAGE_NAME_KEY));
		this.messageType = (String) SerializationConvenience.untokenizeObject(token.getItem(MESSAGE_TYPE_KEY));
		this.version = (String) SerializationConvenience.untokenizeObject(token.getItem(VERSION_KEY));
	}

	@Override
	public StorageToken saveToToken() {
		StorageToken result = super.saveToToken();

		result.setItem(MESSAGE_NAME_KEY, SerializationConvenience.tokenizeObject(this.messageName));
		result.setItem(MESSAGE_TYPE_KEY, SerializationConvenience.tokenizeObject(this.messageType));
		result.setItem(VERSION_KEY, SerializationConvenience.tokenizeObject(this.version));

		return result;
	}

	public String getMessageType() {
		return messageType;
	}

	public void setMessageType(String messageType) {
		this.messageType = messageType;
	}

	public String getVersion() {
		return version;
	}

	public void setVersion(String version) {
		this.version = version;
	}

	public String getMessageName() {
		return messageName;
	}

	public void setMessageName(String messageName) {
		this.messageName = messageName;
	}

	private boolean evaluateMessageType(BaseMessage msg)
	{
		if(this.messageType.equals(GIFT_MESSAGE_TYPE))
			return msg instanceof GIFTMessage;
		
		else if(this.messageType.equals(SUPERGLU_MESSAGE_TYPE))
			return msg instanceof Message;
		
		else if (this.messageType.equals(VHUMAN_MESSAGE_TYPE))
			return msg instanceof VHMessage;
		
		else if (this.messageType.equals(WILDCARD))
			return true;
		
		else 
			return false;
	}
	
	
	private boolean evaluateMessageVersion(BaseMessage msg)
	{
		//For now just accept wildcards
		if(this.version.equals(WILDCARD))
			return true;
		
		
		return false;
	}
	
	
	private boolean evaluateMessageName(BaseMessage msg)
	{
		if(this.messageName.equals(WILDCARD))
			return true;
		
		if(msg instanceof VHMessage)
			return this.messageName.equals(((VHMessage) msg).getFirstWord());
		
		if(msg instanceof Message)
			return this.messageName.equals(((Message) msg).getVerb());
		
		if(msg instanceof GIFTMessage)
			return this.messageName.equals(((GIFTMessage) msg).getHeader());
		
		return false;
	}
	
	
	public boolean evaluateMessage(BaseMessage msg) {
		
		if(!evaluateMessageType(msg))
			return false;
		
		if(!evaluateMessageVersion(msg))
			return false;
		
		if(!evaluateMessageName(msg))
			return false;
		
		return true;
	}

}

File: python_module/SuperGLU/Core/BlackWhiteListEntry.py
from SuperGLU.Util.Serialization import SuperGlu_Serializable, untokenizeObject, tokenizeObject

"""
 * This class represents a single entry in the black or white lists.
 *
 * The each of these fields can be filled with a value or a wildcard (*)
 *
 * @author auerbach
"""
from SuperGLU.Core.Messaging import GIFTMessage, VHMessage
from SuperGLU.Core.Messaging import Message


class BlackWhiteListEntry(SuperGlu_Serializable):

    MESSAGE_TYPE_KEY = "messageType";
    VERSION_KEY = "version";
    MESSAGE_NAME_KEY = "messageName";

    WILDCARD = "*";

    GIFT_MESSAGE_TYPE = "GIFT";
    SUPERGLU_MESSAGE_TYPE = "SuperGLU";
    VHUMAN_MESSAGE_TYPE = "VHuman";



    def __init__(self, messageType="", version="", messageName=""):
        super(BlackWhiteListEntry, self).__init__()
        self._messageType = messageType
        self._version = version
        self._messageName = messageName


    def __eq__(self, other):
        if super(BlackWhiteListEntry, self).__eq__(self, other) is False:
            return False
        if self._messageName is not other._messageName:
            return False
        if self._messageType is not other._messageType:
            return False
        if self._version is not other._version:
            return False

        return True


    def getMessageType(self):
        return self._messageType

    def getMessageName(self):
        return self._messageName

    def getVersion(self):
        return self._version


    def initializeFromToken(self, token, context=None):
        super(BlackWhiteListEntry, self).initializeFromToken(token, context)
        self._messageName = untokenizeObject(token.get(self.MESSAGE_NAME_KEY, ""))
        self._messageType = untokenizeObject(token.get(self.MESSAGE_TYPE_KEY, ""))
        self._version = untokenizeObject(token.get(self.VERSION_KEY))


    def saveToToken(self):
        token = super(BlackWhiteListEntry, self).saveToToken()
        if self._messageName is not None:
            token[self.MESSAGE_NAME_KEY] = tokenizeObject(self._messageName)

        if self._messageType is not None:
            token[self.MESSAGE_TYPE_KEY] = tokenizeObject(self._messageType)

        if self._version is not None:
            token[self.VERSION_KEY] = tokenizeObject(self._version)

        return token


    def evaluateMessageType(self, msg):
        if self._messageType == self.GIFT_MESSAGE_TYPE:
            return isinstance(msg, GIFTMessage)
        if self._messageType == self.VHUMAN_MESSAGE_TYPE:
            return isinstance(msg, VHMessage)
        if self._messageType == self.SUPERGLU_MESSAGE_TYPE:
            return isinstance(msg, Message)
        if self._messageType == self.WILDCARD:
            return True

        return False

    def evaluateMessageVersion(self, msg):
        #for now just accept wildcards until all messages have versions.
        if self._version == self.WILDCARD:
            return True

        return False


    def evaluateMessageName(self, msg):
        if self._messageName == self.WILDCARD:
            return True

        if isinstance(msg, VHMessage):
            return self._messageName == msg.getFirstWord()

        if isinstance(msg, GIFTMessage):
            return self._messageName == msg.getHeader()

        if isinstance(msg, Message):
            return self._messageName == msg.getVerb()

        return False


    def evaluateMessage(self, msg):

        if not self.evaluateMessageType(msg):
            return False

        if not self.evaluateMessageName(msg):
            return False

        if not self.evaluateMessageVersion(msg):
            return False

        return True




class GatewayBlackWhiteListConfiguration(SuperGlu_Serializable):

    ALL_DESTINATIONS = "all"
    EXTERNAL_DESTINATIONS = "external"

    CONFIG_KEY = "config"


    def __init__(self, configDictionary = {}):
        super(GatewayBlackWhiteListConfiguration, self).__init__()
        self._configDictionary = configDictionary

    def addMessage(self, destination, message):
        if destination not in self._configDictionary:
            self._configDictionary[destination] = []

        messageList = self._configDictionary[destination]
        messageList.append(message)
        self._configDictionary[destination] = messageList


    def getMessageList(self, destination):
        return self._configDictionary[destination]


    def getKeys(self):
        return self._configDictionary.keys()



    def initializeFromToken(self, token, context=None):
        super(GatewayBlackWhiteListConfiguration, self).initializeFromToken(token, context)
        self._configDictionary = untokenizeObject(token.get(self.CONFIG_KEY, {}))


    def saveToToken(self):
        token = super(GatewayBlackWhiteListConfiguration, self).saveToToken()

        if self._configDictionary is not None:
            token[self.CONFIG_KEY] = tokenizeObject(self._configDictionary)

        return token
